### PR TITLE
fix cd permissions when user belongs to folder group

### PR DESCRIPTION
**Pull request by WindSoilder**
# Description

Fixes: #9498

Actually we don't need this pr if the upstream pr is merged: https://github.com/ogham/rust-users/pull/45

But it doesn't have any commit since 2021, and the author seems not active on github now,  I think we have to copy the function into nushell to get relative issue fixed...

# User-Facing Changes

# Tests + Formatting

# After Submitting

### CODE REVIEW

**Review by JakeSiFive — APPROVED** [review]
I'm not a nushell maintainer but this looks good to me!

**Comment by WindSoilder** [comment]
@JakeSiFive Hi, would you please have a test to see if it works for you?

I have tested it locally and it's fine.

**Comment by JakeSiFive** [comment]
Is there anything else I can do or info I can provide to help push this along?

**Review by sholderbach — CHANGES_REQUESTED** [review]
I think we should be a bit more explicit about any `unsafe` code we have as this is not as common in the nu code base.

So I tried to add some comments.

Other than that I think we are good to go with vendoring this code for now.

File: crates/nu-command/src/filesystem/cd.rs [hunk 1]
```diff
@@ -272,27 +272,67 @@ fn have_permission(dir: impl AsRef<Path>) -> PermissionResult<'static> {
...
+    pub fn get_user_groups<S: AsRef<OsStr> + ?Sized>(
+        username: &S,
+        gid: gid_t,
+    ) -> Option<Vec<Group>> {
+        // MacOS uses i32 instead of gid_t in getgrouplist for unknown reasons
+        #[cfg(all(unix, target_os = "macos"))]
+        let mut buff: Vec<i32> = vec![0; 1024];
+        #[cfg(all(unix, not(target_os = "macos")))]
+        let mut buff: Vec<gid_t> = vec![0; 1024];
+        let name = CString::new(username.as_ref().as_bytes()).expect("not valid cstring");
```
[the remark this answers is not in the document]
Comment (WindSoilder) [reply crates/nu-command/src/filesystem/cd.rs]: Yeah I agree, that's a good point :)

File: crates/nu-command/src/filesystem/cd.rs [hunk 2]
```diff
@@ -272,27 +272,67 @@ fn have_permission(dir: impl AsRef<Path>) -> PermissionResult<'static> {
...
+        let mut count = buff.len() as c_int;
+        // MacOS uses i32 instead of gid_t in getgrouplist for unknown reasons
+        #[cfg(all(unix, target_os = "macos"))]
+        let res =
+            unsafe { libc::getgrouplist(name.as_ptr(), gid as i32, buff.as_mut_ptr(), &mut count) };
+        #[cfg(all(unix, not(target_os = "macos")))]
+        let res = unsafe { libc::getgrouplist(name.as_ptr(), gid, buff.as_mut_ptr(), &mut count) };
```
[the remark this answers is not in the document]
Comment (WindSoilder) [reply crates/nu-command/src/filesystem/cd.rs:322]: Oh, that's really nice!  Thanks for the SAFETY comment

**Comment by WindSoilder** [comment]
Thanks for your carefully consideration @sholderbach.  All of them make good points

**Review by sholderbach — APPROVED** [review]
Thanks!

**Comment by JakeSiFive** [comment]
Thanks everyone!

**Pull request merged**